dbms/src/Flash/Mpp/MPPTunnel.h

```diff
@@ -44,45 +44,164 @@ namespace DB
 {
 namespace tests
 {
-class MPPTunnelTest;
-class TestMPPTunnelBase;
+class TestMPPTunnel;
 } // namespace tests
 
 class EstablishCallData;
 
+enum class TunnelSenderMode
+{
+    SYNC_GRPC, // Using sync grpc writer
+    LOCAL, // Expose internal memory access, no grpc writer needed
+    ASYNC_GRPC // Using async grpc writer
+};
+
+/// TunnelSender is responsible for consuming data from Tunnel's internal send_queue and do the actual sending work
+/// After TunnelSend finished its work, either normally or abnormally, set ConsumerState to inform Tunnel
+class TunnelSender : private boost::noncopyable
+{
+public:
+    using MPPDataPacketPtr = std::shared_ptr<mpp::MPPDataPacket>;
+    using DataPacketMPMCQueuePtr = std::shared_ptr<MPMCQueue<MPPDataPacketPtr>>;
+    virtual ~TunnelSender() = default;
+    TunnelSender(TunnelSenderMode mode_, DataPacketMPMCQueuePtr send_queue_, PacketWriter * writer_, const LoggerPtr log_, const String & tunnel_id_)
+        : mode(mode_)
+        , send_queue(send_queue_)
+        , writer(writer_)
+        , log(log_)
+        , tunnel_id(tunnel_id_)
+    {
+    }
+    DataPacketMPMCQueuePtr getSendQueue()
+    {
+        return send_queue;
+    }
+    void consumerFinish(const String & err_msg);
+    String getConsumerFinishMsg()
+    {
+        return consumer_state.getMsg();
+    }
+    bool isConsumerFinished()
+    {
+        return consumer_state.msgHasSet();
+    }
+    const LoggerPtr & getLogger() const { return log; }
+    String getTunnelId()
+    {
+        return tunnel_id;
+    }
+
+protected:
+    /// TunnelSender use consumer state to inform tunnel that whether sender has finished its work
+    class ConsumerState
+    {
+    public:
+        ConsumerState()
+            : future(promise.get_future())
+        {
+        }
+
+        // before finished, must be called without protection of mu
+        String getMsg()
+        {
+            future.wait();
+            return future.get();
+        }
+        void setMsg(const String & msg)
+        {
+            promise.set_value(msg);
```

Comment:
Not releated to this pr, but do we need some protections to make sure `promise.set_value` is not called multiple times? Since [here](https://en.cppreference.com/w/cpp/thread/promise/set_value) says "An exception is thrown if there is no shared state or the shared state already stores a value or exception."

Reply:
Currently, the only way to set consumer_state's msg is to call consumerFinish. In consumerFinish, there is the protection, check if msg already set, if so, just do nothing and return.

Reply:
How about we enhance `setMsg` as 
```
void setMsg(const String & msg)
{
    bool old_value = false;
    if (!msg_has_set.compare_exchange_strong(old_val, true, std::memory_order_seq_cst, std::memory_order_relaxed))
        return;
    promise.set_value(msg);
}
```
And `consumerFinish` don't need to involve lock.

Reply:
The tricky part of this is that, it's an obvious bug if `promise.set_value` is called more than once. Do we need to protect for potential bug and pretend nothing happened, or just leave it throw then we could quickly know there's a bug?

Reply:
Done